jobsdb/unionQuery.go

```diff
@@ -0,0 +1,284 @@
+	}
+
+	if len(parameterFilters) > 0 {
+		sqlStatement += " AND " + constructParameterJSONQuery("jobs", parameterFilters)
+	}
+
+	//avoinding AfterJobID for now
+
+	if params.UseTimeFilter {
+		sqlStatement += fmt.Sprintf(" AND created_at < %s", params.Before)
+	}
+
+	if order {
+		sqlStatement += " ORDER BY jobs.job_id"
+	}
+
+	if count > 0 {
+		sqlStatement += fmt.Sprintf(" LIMIT %d", count)
+	}
+
+	if params.EventCount > 0 {
+		sqlStatement = fmt.Sprintf(`SELECT * FROM (`+sqlStatement+`) AS subquery WHERE running_event_counts - event_count + 1 <= %d;`, params.EventCount)
+	}
+
+	return sqlStatement
+}
+
+func (jd *HandleT) GetUnprocessedUnion(unionMap map[string]int, params GetQueryParamsT) []*JobT {
+	// if params.JobCount == 0 { //how do we use this check..? is it required anymore
+	// 	return []*JobT{}
+	// }
+
+	//add stats
+
+	//The order of lock is very important. The migrateDSLoop
+	//takes lock in this order so reversing this will cause
+	//deadlocks
+	jd.dsMigrationLock.RLock()
+	jd.dsListLock.RLock()
+	defer jd.dsMigrationLock.RUnlock()
+	defer jd.dsListLock.RUnlock()
+
+	dsList := jd.getDSList(false)
+	outJobs := make([]*JobT, 0)
+
+	//removed count assert, because that params.count is not used..?
+
+	for _, ds := range dsList {
+		jobs := jd.getUnprocessedUnionDS(ds, unionMap, params)
+		outJobs = append(outJobs, jobs...)
+		// count -= len(jobs) //this is not necessary anymore right?
+
+		//we could further limit jobs by eventCount - ignoring that for now
+		// just focusing on getting so many rows of jobs for now
+	}
+
+	return outJobs
+}
+
+func (jd *HandleT) getUnprocessedUnionDS(ds dataSetT, unionMap map[string]int, params GetQueryParamsT) []*JobT {
+	queryString := jd.getUnprocessedUnionQuerystring(unionMap, ds, params)
+
+	var rows *sql.Rows
+	var err error
+
+	rows, err = jd.dbHandle.Query(queryString)
+	jd.assertError(err)
+
+	defer rows.Close()
+
+	var jobList []*JobT
+	for rows.Next() {
+		var job JobT
+		var _null int
+		err := rows.Scan(&job.JobID, &job.UUID, &job.UserID, &job.Parameters, &job.CustomVal,
+			&job.EventPayload, &job.EventCount, &job.CreatedAt, &job.ExpireAt, &job.Customer, &_null)
+		jd.assertError(err)
+		jobList = append(jobList, &job)
+
+		unionMap[job.Customer] -= 1
+		if unionMap[job.Customer] == 0 {
+			delete(unionMap, job.Customer)
+		}
+	}
+
+	//do cache stuff here
+
+	return jobList
+}
+
+//Processed
+
+func (jd *HandleT) GetProcessedUnion(unionMap map[string]int, params GetQueryParamsT) []*JobT {
+
+	//The order of lock is very important. The migrateDSLoop
+	//takes lock in this order so reversing this will cause
+	//deadlocks
+	jd.dsMigrationLock.RLock()
+	jd.dsListLock.RLock()
+	defer jd.dsMigrationLock.RUnlock()
+	defer jd.dsListLock.RUnlock()
+
+	dsList := jd.getDSList(false)
+	outJobs := make([]*JobT, 0)
+
+	for _, ds := range dsList {
+		jobs := jd.getProcessedUnionDS(ds, unionMap, params)
+		outJobs = append(outJobs, jobs...)
```

Comment:
Similar to `GetUnprocessedUnion`, we should check and stop early if all counters are zero. Maybe also assert for negative counts.

Reply:
Please elaborate more about the negative counts, maybe with an example showing how we might come across it.